### PR TITLE
Fix overconstrained Send impls in btree internals

### LINKED ISSUE #215: `higher-ranked lifetime error` with a simple example involving a BTreeMap
The following ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e3fbe7f1f693f17c7daa8bcf4cf31d9e))
```
#[async_trait::async_trait]
trait T {
    async fn f(&self);
}
#[async_trait::async_trait]
// Compiles when the BTreeMap is replaced by a HashMap
impl T for BTreeMap<u32, Box<dyn Send + Sync + 'static>> {
    async fn f(&self) {
        for a in self {
            test().await;
        }
    }
}
async fn test() {}
```
results in
```
error: higher-ranked lifetime error
  --> src/lib.rs:11:23
   |
11 |       async fn f(&self) {
   |  _______________________^
12 | |         for a in self {
13 | |             test().await;
14 | |         }
15 | |     }
   | |_____^
   |
   = note: could not prove `for<'r, 's, 't0, 't1, 't2> Pin<Box<impl for<'r, 's, 't0, 't1> Future<Output = ()>>>: CoerceUnsized<Pin<Box<(dyn Future<Output = ()> + Send + 't2)>>>`
```
however, it works fine when the BTreeMap is replaced by a HashMap.
Might that be an issue with `async_trait`, or something I misunderstood?

### PR DESCRIPTION
**Pull request by dtolnay**
Fixes https://github.com/dtolnay/async-trait/issues/215.

Minimal repro:

```rust
use std::collections::btree_map::Iter;

fn require_send<T: Send>(_: T) {}

fn main() {
    require_send(async {
        let _iter = None::<Iter<(), &()>>;
        async {}.await;
    });
}
```

```console
error: higher-ranked lifetime error
 --> src/main.rs:6:5
  |
6 | /     require_send(async {
7 | |         let _iter = None::<Iter<(), &()>>;
8 | |         async {}.await;
9 | |     });
  | |______^
  |
  = note: could not prove `impl Future<Output = ()>: Send`
```

Not-quite-so-minimal repro:

```rust
use std::collections::BTreeMap;
use std::future::Future;

fn spawn<T: Future + Send>(_: T) {}

async fn f() {
    let map = BTreeMap::<u32, Box<dyn Send + Sync>>::new();
    for _ in &map {
        async {}.await;
    }
}

fn main() {
    spawn(f());
}
```

```console
error: higher-ranked lifetime error
  --> src/main.rs:14:5
   |
14 |     spawn(f());
   |     ^^^^^^^^^^
   |
   = note: could not prove `impl Future<Output = ()>: Send`
```

I am not familiar with the btree internals, but it seems clear to me that the `async fn f` above should return a Send future. Using HashMap instead of BTreeMap in that code makes it already return a Send future.

The _"higher-ranked lifetime error"_ message may be a regression in Rust 1.63. Using older compilers the error message was more detailed:

```console
error: implementation of `Send` is not general enough
  --> src/main.rs:14:5
   |
14 |     spawn(f());
   |     ^^^^^ implementation of `Send` is not general enough
   |
   = note: `Send` would have to be implemented for the type `alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Immut<'0>, u32, Box<(dyn Send + Sync + '1)>, alloc::collections::btree::node::marker::LeafOrInternal>`, for any two lifetimes `'0` and `'1`...
   = note: ...but `Send` is actually implemented for the type `alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Immut<'2>, u32, Box<dyn Send + Sync>, alloc::collections::btree::node::marker::LeafOrInternal>`, for some specific lifetime `'2`

error: implementation of `Send` is not general enough
  --> src/main.rs:14:5
   |
14 |     spawn(f());
   |     ^^^^^ implementation of `Send` is not general enough
   |
   = note: `Send` would have to be implemented for the type `alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Immut<'0>, u32, Box<(dyn Send + Sync + '1)>, alloc::collections::btree::node::marker::Leaf>`, for any two lifetimes `'0` and `'1`...
   = note: ...but `Send` is actually implemented for the type `alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Immut<'2>, u32, Box<dyn Send + Sync>, alloc::collections::btree::node::marker::Leaf>`, for some specific lifetime `'2`
```

### CODE REVIEW

**Comment by Aaron1011** [comment]
The general issue is https://github.com/rust-lang/rust/issues/64552 - this fixes a particular instance by eliminating the need for rustc to attempt to prove a lifetime bound on the erased witness types.

As long as the new impl is sound, I think it's fine to merge this. However, users will continue to see these unhelpful errors in other cases until the underlying issue is resolved.

File: src/test/ui/btreemap/btreemap_autotraits_without_lifetime.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// check-fail
```
Comment (thomcc) [inline src/test/ui/btreemap/btreemap_autotraits_without_lifetime.rs:1]: I think most people don't bother to run the uitests locally for library patches, and the error message is the same as any other Sync error, so IMO its a bit better to do this as a `compile_fail` doctest on something private, and put `check-pass` in the normal tests.

That said this is unlikely to unexpectedly break in the future, so if you feel strongly that they should be a `uitest`, fine by me as-is.

**Review by thomcc — APPROVED** [review]
Huh, good catch. It doesn't make sense that we'd bound those on the same lifetime we use in the marker, so this looks fine to me.

I don't think this needs libs-api sign-off, as it's hard to imagine how this could be intentionally forbidden, but you'd have a better idea (as someone actually part of that team) so I'll leave that choice to you.

r=me if you move the tests out of uitest (or feel strongly that they shouldn't be moved).

**Comment by dtolnay** [comment]
Moved test to library/alloc/tests/autotraits.rs and extended it to cover alloc::collections more thoroughly, which revealed a `linked_list::DrainFilter` bug https://github.com/rust-lang/rust/issues/43244#issuecomment-1268857915.

I left out negative tests for now but I am thinking of following up with some based on https://github.com/dtolnay/proc-macro2/blob/1.0.46/tests/marker.rs or the static_assertions crate.

**Pull request merged**